File: common/src/main/java/dev/obscuria/fragmentum/util/signal/Signal4.java
package dev.obscuria.fragmentum.util.signal;

@SuppressWarnings("unused")
public final class Signal4<P1, P2, P3, P4> extends Signal<Signal4.Listener<P1, P2, P3, P4>>
{
    public void emit(P1 p1, P2 p2, P3 p3, P4 p4)
    {
        emit(it -> it.consume(p1, p2, p3, p4));
    }

    @FunctionalInterface
    public interface Listener<P1, P2, P3, P4>
    {
        void consume(P1 p1, P2 p2, P3 p3, P4 p4);
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal4.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.tools.signal

class Signal4<P1, P2, P3, P4> : Signal<Signal4.Listener<P1, P2, P3, P4>>() {

    fun emit(p1: P1, p2: P2, p3: P3, p4: P4): Unit = error("API stub")

    fun interface Listener<P1, P2, P3, P4> {

        fun consume(p1: P1, p2: P2, p3: P3, p4: P4)
    }
}